### PR TITLE
phpstan: Add many property declarations and re-enable undefined property error

### PR DESCRIPTION
**Pull request by bpfoley**
This is somewhat error prone because I tried to make the properties `protected` or `private` where I could, and had to try to figure out the correct type hints.

PHPStan itself caught a lot of cases where I got this wrong, but there were other places where it failed because many of our functions that return objects don't have `@return` annotations, so PHPStan didn't know the types of things like the following, and couldn't see the property accesses though it.

```
$ppage = get_requested_PPage($_POST); // function has no @return, so PHPStan thinks $ppage's type is 'object'.
...
... = $ppage->lpage
```

Another common failure was that most of our arrays aren't annotated with their element types so PHPStan couldn't detect property access of array elements via iterators.

For both these scenarios, I've sprinkled more annotations, but we'll have to add a lot more to get full coverage.

There is a sandbox at https://www.pgdp.org/~bfoley/c.branch/property-is-theft/ and I have done some basic testing of many of the pages on the site.

### CODE REVIEW

File: accounts/activate.php [hunk 1]
```diff
@@ -71,6 +71,8 @@
 }
 
 // Verify we can create the user's forum account, and bail if we can't.
+/** @var NonactivatedUser */
+$user = $user;
```
Comment (cpeel) [inline accounts/activate.php:75]: What's going on here?

**Review by cpeel — COMMENTED** [review]
I want to give this some testing of my own when I get back from vacation but the changes look good inasmuch as I can remember private/protected/public access.

File: accounts/activate.php (hunk 1 above)
Comment (bpfoley) [reply accounts/activate.php:75]: At the time it was the only way I could think of telling PHPStan that `$user` is in fact a `NonactivatedUser`.

Now, I realize that it's probably an incorrect PHPDoc annotation on `load_from_id`. That function has the docs 
```php
/**
  * Load a NonactivatedUser record by ID
  * ...
  * @return object NonactivatedUser
  */
  public static function load_from_id($id) { ... }
```

the `@return object NonactivatedUser` should be `@return NonactivatedUser`

Comment (bpfoley) [reply accounts/activate.php:75]: Actually, no, I'm wrong. I'd fixed up the `@return NonactivatedUser` already. The actual problem is that PHPStan interprets the fact that the only initialization of `$user` occurs in a fallible `try` block, so it interprets the type as being as generic as possible.

```
  try {
      $user = NonactivatedUser::load_from_id($reg_token);
  } catch (NonexistentNonactivatedUserException $exception) { ... }
 ```

The workaround is to explicitly assign `$user = null;` in the `catch`. Which, logically means the type of `$user` should be `NonactiveUser|null` after the try/catch block, but PHPStan silently ignores the `|null` part in this context.

**Comment by bpfoley** [comment]
> I want to give this some testing of my own when I get back from vacation but the changes look good inasmuch as I can remember private/protected/public access.

Great. If you remember, this is what motivated me to work on #1130. I'll try to finish off the smoketests, and if we get those committed first they should give us a bit more confidence that we haven't inadvertently introduced runtime errors with overly strict annotations.

File: pinc/page_tally.inc [hunk 2]
```diff
@@ -120,7 +120,8 @@ function user_get_ELR_page_tally($username)
 
 /**
  * Get the user's page tally neighborhood
- *
+ * @param string $tally_name
+ * @param string $username
  * @param int $radius
  *   (maximum) number of neighbors (on each side) to include in
  *   the neighborhood. (It will include fewer that the maximum iff the target
```
[the remark this answers is not in the document]
Comment (bpfoley) [reply pinc/page_tally.inc:126]: Done.

File: pinc/Round.inc [hunk 3]
```diff
@@ -42,10 +42,38 @@ class Rounds extends Stages
 // A container for various constants relating to a particular round of proofreading.
 class Round extends Stage
 {
+    public int $round_number;
+    /** @var array<string, string[]> */
```
[the remark this answers is not in the document]
Comment (bpfoley) [reply pinc/Round.inc:46]: It's a PHPDoc syntax to say what the types of the keys and values in an associative array are. In this case, I'm saying that `$pi_tools` is an associative array whose keys are strings, and the values are lists of strings. The doc for the constructor shows an example of this.

```
[
    'popup_links' => [ 'link1', 'link2', ... ],
    'tool_buttons' => [ 'button1', 'button2', ... ],
    'tool_links' => [ 'link1', 'link2', ... ],
]
``` 

As you say, the issue of PHP implicitly casting "int-y" strings to numbers is a real one, as are PHPStan's loosened checks to accommodate this. But still, having *some* type hints on the array is better than none at all.

[the remark this answers is not in the document]
Comment (srjfoo) [reply pinc/Round.inc:46]: I definitely agree that it's better to have hints than not. I just wondered how that would affect what we're doing. Remember, I'm a complete noob at this.

**Comment by cpeel** [comment]
Apologies for spacing on this, Brian. For some reason I'd forgotten you were waiting on me.

I'd like for us to break this into two parts, please. The first one is the bulk of the commits, specifically all of the following:
```
    phpstan.neon: List all classes without explicitly defined properties
    Consistently treat Project::archived as a bool, not an int or string
    CharSuites.inc: Declare PickerSet, CharSuite properties
    CharacterSelector.inc: Declare properties
    faq_central.php: Add class property declarations
    SpecialDays: Add class property declarations.
    ProofreadingToolbox.inc: Declare class properties
    Quiz.inc: Declare class properties
    BadWordAccumulator: Declare class properties
    page_compare.php: Declare class properties
    ImageSource: Declare properties, less string/int punning
    SortUtility: Declare properties; fix types/name typo
    Stopwatch: Declare class properties
    POFile: Add dynamic property annotations
    MARCRecord: Annotate class with all dynamic properties
    ThemedTable: Add class property declarations
    ProjectState: Declare class properties
    Fix PHPDoc annotations to '@return Foo' not '@return object Foo'
    *User* classes: Annotate with dynamic properties
    tasks.php: Quell phpstan property access warning
```

Those cover some that are simple or "low" impact if we get a public/private/protected entry wrong. That leaves ~6ish that I want to take a more focused look at and exercise some weird use-cases. I'm happy to get the above in as-is (with the usual merge conflict resolution caveats) ASAP.

**Comment by bpfoley** [comment]
Closing this and continuing the work in #1189

**Pull request closed without merging**